src/main/perf/DocGrouper.java

```diff
@@ -0,0 +1,251 @@
+package perf;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import static org.apache.lucene.tests.util.TestUtil.randomRealisticUnicodeString;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Consuming {@link perf.LineFileDocs.LineFileDoc}, group them and put the grouped docs into
+ * a thread-safe queue.
+ */
+public abstract class DocGrouper {
```

Comment:
Can we make this final, non-abstract, and rename it to `TextDocGrouper` maybe?  No need for separate subclass since we have no binary case anymore?

Reply:
We still need a `NoGroupImpl` to deal with the case where we don't need groups, or we can handle that difference in `LineFileDocs` which requires several if/else there... I'm ok with either but think this might be (slightly) cleaner?